test/blackbox/PubSubReader.hpp

```diff
@@ -840,24 +842,50 @@ class PubSubReader
         type data;
         eprosima::fastrtps::SampleInfo_t info;
 
-        if(subscriber->takeNextData((void*)&data, &info))
+        if (take_)
         {
-            returnedValue = true;
+            if(subscriber->takeNextData((void*)&data, &info))
+            {
+                returnedValue = true;
 
-            std::unique_lock<std::mutex> lock(mutex_);
+                std::unique_lock<std::mutex> lock(mutex_);
 
-            // Check order of changes.
-            ASSERT_LT(last_seq, info.sample_identity.sequence_number());
-            last_seq = info.sample_identity.sequence_number();
+                // Check order of changes.
+                ASSERT_LT(last_seq, info.sample_identity.sequence_number());
+                last_seq = info.sample_identity.sequence_number();
 
-            if(info.sampleKind == eprosima::fastrtps::rtps::ALIVE)
+                if(info.sampleKind == eprosima::fastrtps::rtps::ALIVE)
+                {
+                    auto it = std::find(total_msgs_.begin(), total_msgs_.end(), data);
+                    ASSERT_NE(it, total_msgs_.end());
+                    total_msgs_.erase(it);
+                    ++current_received_count_;
+                    default_receive_print<type>(data);
+                    cv_.notify_one();
+                }
+            }
+        }
+        else
+        {
+            if(subscriber->readNextData((void*)&data, &info))
             {
-                auto it = std::find(total_msgs_.begin(), total_msgs_.end(), data);
-                ASSERT_NE(it, total_msgs_.end());
-                total_msgs_.erase(it);
-                ++current_received_count_;
-                default_receive_print<type>(data);
-                cv_.notify_one();
+                returnedValue = true;
+
+                std::unique_lock<std::mutex> lock(mutex_);
+
+                // Check order of changes.
+                ASSERT_LT(last_seq, info.sample_identity.sequence_number());
+                last_seq = info.sample_identity.sequence_number();
+
+                if(info.sampleKind == eprosima::fastrtps::rtps::ALIVE)
+                {
+                    auto it = std::find(total_msgs_.begin(), total_msgs_.end(), data);
+                    ASSERT_NE(it, total_msgs_.end());
+                    total_msgs_.erase(it);
+                    ++current_received_count_;
+                    default_receive_print<type>(data);
+                    cv_.notify_one();
+                }
```

Comment:
The only difference between both branches is [`take`|`read`]`NextData`, right?
They could be merged to avoid code duplication in somethin like:

    bool success = take_ ? takeNextData... : readNextData...;
    if (success)
    {
        ...
    }

Reply:
Cool, changed